File: demo/node/rntuple_test.js
import { version, openFile, TSelector } from 'jsroot';

import { readHeaderFooter, rntupleProcess } from 'jsroot/rntuple';

console.log(`JSROOT version ${version}`);


let filename = './rntuple_test.root',
    any_error = false;

if (process?.argv && process.argv[2])
   filename = process.argv[2];

const file = await openFile(filename),
      rntuple = await file.readObject('Data'),
      builder = await readHeaderFooter(rntuple);

console.log('Performing Validations and debugging info');

if (builder?.name !== 'Data')
   console.error('FAILURE: name differs from expected');
else
   console.log('OK: name is', builder?.name);


if (builder.description !== '')
   console.error('FAILURE: description should be the empty string');
else
   console.log('OK: description is empty string');

if (builder.xxhash3 === undefined || builder.xxhash3 === null)
   console.warn('WARNING: xxhash3 is missing');
else
   console.log('OK: xxhash3 is', '0x' + builder.xxhash3.toString(16).padStart(16, '0'));

// Fields Check

if (!builder?.fieldDescriptors?.length)
   console.error('FAILURE: No fields deserialized');
else {
   console.log(`OK: ${builder.fieldDescriptors.length} field(s) deserialized`);
   for (let i = 0; i < builder.fieldDescriptors.length; ++i) {
      const field = builder.fieldDescriptors[i];
      if (!field.fieldName || !field.typeName)
         console.error(`FAILURE: Field ${i} is missing name or type`);
      else
         console.log(`OK: Field ${i}: ${field.fieldName} (${field.typeName})`);
      if (i === 0) {
         if (field.fieldName !== 'IntField' || field.typeName !== 'std::int32_t') {
            any_error = true;
            console.error(`FAILURE: First field should be 'IntField (std::int32_t)' but got '${field.fieldName} (${field.typeName})'`);
         }
      } else if (i === builder.fieldDescriptors.length - 1) {
         if (field.fieldName !== '_1' || field.typeName !== 'bool') {
            any_error = true;
            console.error(`FAILURE: Last field should be '_1 (bool)' but got '${field.fieldName} (${field.typeName})'`);
         }
      }
   }
}

// Column Check

if (!rntuple.builder?.columnDescriptors?.length) {
   any_error = true;
   console.error('FAILURE: No columns deserialized');
} else {
   console.log(`OK: ${builder.columnDescriptors.length} column(s) deserialized`);
   for (let i = 0; i < builder.columnDescriptors.length; ++i) {
      const column = builder.columnDescriptors[i];
      if (column.fieldId === undefined || column.fieldId === null)
         console.error(`FAILURE: Column ${i} is missing fieldId`);
      else
         console.log(`OK: Column ${i} fieldId: ${column.fieldId} `);
      if ((i === 0) && (column.fieldId !== 0)) {
         any_error = true;
         console.error('FAILURE: First column should be for fieldId 0');
      } else if ((i === builder.columnDescriptors.length - 1) && (column.fieldId !== builder.fieldDescriptors.length - 1)) {
         any_error = true;
         console.error(`FAILURE: Last column should be for fieldId ${builder.fieldDescriptors.length - 1}`);
      }
   }
}

// Setup selector to process all fields (so cluster gets loaded)
const selector = new TSelector(),
      fields = ['IntField', 'FloatField', 'DoubleField',
                'Float16Field', 'Real32Trunc', 'Real32Quant',
                'StringField', 'BoolField',
                'BitsetField',
                'ArrayInt', 'VariantField', 'TupleField',
                'VectString', 'VectInt', 'VectBool', 'Vect2Float', 'Vect2Bool', 'MultisetField',
                'MapStringFloat', 'MapIntDouble', 'MapStringBool' ],
      epsilonValues = { Real32Trunc: 0.3, Real32Quant: 1e-4, Float16Field: 1e-2 };

for (const f of fields)
   selector.addBranch(f);

selector.Begin = () => {
   console.log('\nBegin processing to load cluster data...');
};


// Now validate entry data
const EPSILON = 1e-7;

function compare(expected, value, eps) {
   if (typeof expected === 'number')
      return Math.abs(value - expected) < (eps ?? EPSILON);
   if (typeof expected === 'object') {
      if (expected.length !== undefined) {
         if (expected.length !== value.length)
            return false;
         for (let j = 0; j < expected.length; ++j) {
            if (!compare(expected[j], value[j]))
               return false;
         }
      } else {
         for (const key in expected) {
            if (!compare(expected[key], value[key]))
               return false;
         }
      }
      return true;
   }
   return expected === value;
}

selector.Process = function(entryIndex) {
   console.log(`\nChecking entry ${entryIndex}:`);

   const expectedValues = {
      IntField: entryIndex,
      FloatField: entryIndex * entryIndex,
      DoubleField: entryIndex * 0.5,
      Float16Field: entryIndex * 0.1987333,
      Real32Trunc: 123.45 * entryIndex,
      Real32Quant: 0.03 * (entryIndex % 30),
      StringField: `entry_${entryIndex}`,
      BoolField: entryIndex % 3 === 1,
      ArrayInt: [entryIndex + 1, entryIndex + 2, entryIndex + 3, entryIndex + 4, entryIndex + 5],
      VariantField: null,
      TupleField: { _0: `tuple_${entryIndex}`, _1: entryIndex*3, _2: (entryIndex % 3 === 1) },
      VectString: [],
      VectInt: [],
      VectBool: [],
      Vect2Float: [],
      Vect2Bool: [],
      MultisetField: [],
      MapStringFloat: [],
      MapIntDouble: [],
      MapStringBool: [],
      BitsetField: 1 << entryIndex * 3 % 25
   }, npx = (entryIndex + 5) % 7;

   switch (entryIndex % 3) {
      case 0: expectedValues.VariantField = `varint_${entryIndex}`; break;
      case 1: expectedValues.VariantField = entryIndex; break;
      case 2: expectedValues.VariantField = (entryIndex % 2 === 0); break;
   }

   for (let j = 0; j < npx; ++j) {
      expectedValues.VectString.push(`str_${j}`);
      expectedValues.VectInt.push(-j);
      expectedValues.VectBool.push(j % 2 === 1);

      expectedValues.MultisetField.push(`multiset_${j % 3}`);

      expectedValues.MapStringFloat.push({ first: `key_${j}`, second: j * 7 });
      expectedValues.MapIntDouble.push({ first: j * 11, second: j * 0.2 });
      expectedValues.MapStringBool.push({ first: `bool_${j}`, second: j % 3 === 0 });

      const npy = 1 + entryIndex % 3, vf = [], vb = [];
      for (let k = 0; k < npy; ++k) {
         vf.push(k * 1.1);
         vb.push(k % 2 === 0);
      }
      expectedValues.Vect2Float.push(vf);
      expectedValues.Vect2Bool.push(vb);
   }

   expectedValues.MultisetField.sort();

   for (const field of fields) {
      try {
         const value = this.tgtobj[field],
               expected = expectedValues[field];

         if (!compare(expected, value, epsilonValues[field])) {
            console.error(`FAILURE: ${field} at entry ${entryIndex} expected ${JSON.stringify(expected)}, got ${JSON.stringify(value)}`);
            any_error = true;
         } else
            console.log(`OK: ${field} at entry ${entryIndex} = ${JSON.stringify(value)}`);
      } catch (err) {
         console.error(`ERROR: Failed to read ${field} at entry ${entryIndex}: ${err.message}`);
         any_error = true;
      }
   }
};

selector.Terminate = () => {
   console.log('Finished processing');
};

// Run rntupleProcess to ensure cluster is loaded
await rntupleProcess(rntuple, selector);

// run again, but select only two entries
await rntupleProcess(rntuple, selector, { elist: [3, 7] });

if (any_error)
   console.error('\nFAILURE when verifying file content');
else
   console.log('\nTest OK');

File: demo/node/rntuple_test.cxx
#include <ROOT/RNTupleModel.hxx>
#include <ROOT/RNTupleWriter.hxx>


#include <cstdio>
#include <iostream>
#include <memory>
#include <vector>
#include <array>
#include <utility>
#include <string>
#include <variant>
#include <tuple>
#include <map>
#include <set>
#include <bitset>

class TestClass {
   public:
      std::string fName;
      std::string fTitle;
      double fValue{0.};

};

#ifdef __ROOTCLING__
#pragma link C++ class std::map<std::string,float>+;
#pragma link C++ class std::map<int,double>+;
#pragma link C++ class std::map<std::string,bool>+;
#pragma link C++ class std::multiset<std::string>+;
#pragma link C++ class std::variant<std::string,int,bool>+;
#pragma link C++ class std::tuple<std::string,int,bool>+;
#pragma link C++ class std::bitset<25>+;
#pragma link C++ class std::bitset<117>+;
#pragma link C++ class TestClass+;
#endif


// Where to store the ntuple of this example
constexpr char const* kNTupleFileName = "rntuple_test.root";

// Number of events to generate
constexpr int kNEvents = 10;

// Generate kNEvents with vectors in kNTupleFileName
void rntuple_test()
{
   // We create a unique pointer to an empty data model
   auto model = ROOT::RNTupleModel::Create();

   // Creating fields of std::vector is the same as creating fields of simple types.  As a result, we get
   // shared pointers of the given type
   auto IntField = model->MakeField<int>("IntField");
   auto FloatField = model->MakeField<float>("FloatField");
   auto Float16Field = model->MakeField<float>("Float16Field");
   model->GetMutableField("Float16Field").SetColumnRepresentatives({{ROOT::ENTupleColumnType::kReal16}});

   auto Real32Trunc = model->MakeField<float>("Real32Trunc");
   dynamic_cast<ROOT::RRealField<float> &>(model->GetMutableField("Real32Trunc")).SetTruncated(20);

   auto Real32Quant = model->MakeField<float>("Real32Quant");
   dynamic_cast<ROOT::RRealField<float> &>(model->GetMutableField("Real32Quant")).SetQuantized(0., 1., 14);

   auto DoubleField = model->MakeField<double>("DoubleField");
   auto StringField = model->MakeField<std::string>("StringField");
   auto BoolField = model->MakeField<bool>("BoolField");
   auto VariantField = model->MakeField<std::variant<std::string,int,bool>>("VariantField");
   auto TupleField = model->MakeField<std::tuple<std::string,int,bool>>("TupleField");
   auto ArrayInt = model->MakeField<std::array<int,5>>("ArrayInt");
   auto BitsetField = model->MakeField<std::bitset<25>>("BitsetField");
   auto LargeBitsetField = model->MakeField<std::bitset<117>>("LargeBitsetField");
   auto AtomicDoubleField = model->MakeField<std::atomic<double>>("AtomicDoubleField");
   auto TestClassField = model->MakeField<TestClass>("TestClassField");
   auto VectString   = model->MakeField<std::vector<std::string>>("VectString");
   auto VectInt   = model->MakeField<std::vector<int>>("VectInt");
   auto VectBool   = model->MakeField<std::vector<bool>>("VectBool");
   auto MultisetField = model->MakeField<std::multiset<std::string>>("MultisetField");
   auto Vect2Float   = model->MakeField<std::vector<std::vector<float>>>("Vect2Float");
   auto Vect2Bool   = model->MakeField<std::vector<std::vector<bool>>>("Vect2Bool");
   auto MapStringFloat  = model->MakeField<std::map<std::string,float>>("MapStringFloat");
   auto MapIntDouble   = model->MakeField<std::map<int,double>>("MapIntDouble");
   auto MapStringBool  = model->MakeField<std::map<std::string,bool>>("MapStringBool");

   // We hand-over the data model to a newly created ntuple of name "F", stored in kNTupleFileName
   // In return, we get a unique pointer to an ntuple that we can fill
   auto writer = ROOT::RNTupleWriter::Recreate(std::move(model), "Data", kNTupleFileName);

   for (int i = 0; i < kNEvents; i++) {

      *IntField = i;
      *FloatField = i*i;

      *Float16Field = 0.1987333 * i;  // stored as 16 bits float
      *Real32Trunc = 123.45 * i; // here only 20 bits preserved
      *Real32Quant = 0.03 * (i % 30); // value should be inside [0..1]

      *DoubleField = 0.5 * i;
      *StringField = "entry_" + std::to_string(i);
      *BoolField = (i % 3 == 1);
      *ArrayInt = { i + 1, i + 2, i + 3, i + 4, i + 5 };
      switch (i % 3) {
         case 0: *VariantField = std::string("varint_") + std::to_string(i); break;
         case 1: *VariantField = i; break;
         case 2: *VariantField = (bool) (i % 2 == 0); break;
      }

      *TupleField = { std::string("tuple_") + std::to_string(i), i * 3, (i % 3 == 1) };

      TestClassField->fName = "name_" + std::to_string(i);
      TestClassField->fTitle = "title_" + std::to_string(i);
      TestClassField->fValue = i;

      VectString->clear();
      VectInt->clear();
      VectBool->clear();

      MultisetField->clear();

      MapStringFloat->clear();
      MapIntDouble->clear();
      MapStringBool->clear();

      Vect2Float->clear();
      Vect2Bool->clear();

      BitsetField->reset();
      LargeBitsetField->reset();

      BitsetField->set(i * 3 % 25, true);

      LargeBitsetField->set((i + 7) % 117, true);
      LargeBitsetField->set((i + 35) % 117, true);

      *AtomicDoubleField = 111.444 * i;

      int npx = (i + 5) % 7;
      for (int j = 0; j < npx; ++j) {
         VectString->emplace_back("str_" + std::to_string(j));
         VectInt->emplace_back(-j);
         VectBool->emplace_back(j % 2 == 1);

         MapStringFloat->emplace("key_" + std::to_string(j), j*7);
         MapIntDouble->emplace(j*11, j*0.2);
         MapStringBool->emplace("bool_" + std::to_string(j), j % 3 == 0);

         MultisetField->insert("multiset_" + std::to_string(j % 3));

         int npy = 1 + i % 3;
         std::vector<float> vf;
         std::vector<bool> vb;
         for (int k = 0; k < npy; ++k) {
            vf.emplace_back(k*1.1);
            vb.emplace_back(k % 2 == 0);
         }
         Vect2Float->emplace_back(vf);
         Vect2Bool->emplace_back(vb);
      }

      writer->Fill();
    }

   // The ntuple unique pointer goes out of scope here.  On destruction, the ntuple flushes unwritten data to disk
   // and closes the attached ROOT file.
}
